Sources/SwiftInspectorVisitors/NestableTypeVisitor.swift

```diff
@@ -0,0 +1,204 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import SwiftSyntax
+
+/// A visitor capable of parsing (possibly-nested) type information from classes, structs, and enums.
+public final class NestableTypeVisitor: SyntaxVisitor {
+
+  public init(parentType: TypeDescription? = nil) {
+    self.parentType = parentType
+  }
+
+  deinit {
+    assert(!topLevelParsingTracker.isParsing)
+  }
+
+  /// All the classes found by this visitor
+  public var classes: [ClassInfo] {
+    [topLevelDeclaration?.classInfo].compactMap { $0 } + innerClasses
+  }
+  /// All the structs found by this visitor
+  public var structs: [StructInfo] {
+    [topLevelDeclaration?.structInfo].compactMap { $0 } + innerStructs
+  }
+  /// All of the enums found by this visitor.
+  public var enums: [EnumInfo] {
+    [topLevelDeclaration?.enumInfo].compactMap { $0 } + innerEnums
+  }
+
+  public override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
+    visitNestableDeclaration(node: node, topLevelDeclarationCreator: { .topLevelClass($0) })
+  }
+
+  public override func visitPost(_ node: ClassDeclSyntax) {
+    decrementParsingTracker()
+  }
+
+  public override func visit(_ node: StructDeclSyntax) -> SyntaxVisitorContinueKind {
+    visitNestableDeclaration(node: node, topLevelDeclarationCreator: { .topLevelStruct($0) })
+  }
+
+  public override func visitPost(_ node: StructDeclSyntax) {
+    decrementParsingTracker()
+  }
+
+  public override func visit(_ node: EnumDeclSyntax) -> SyntaxVisitorContinueKind {
+    visitNestableDeclaration(node: node, topLevelDeclarationCreator: { .topLevelEnum($0) })
+  }
+
+  public override func visitPost(_ node: EnumDeclSyntax) {
+    decrementParsingTracker()
+  }
+
+  public override func visit(_ node: ProtocolDeclSyntax) -> SyntaxVisitorContinueKind {
+    // We've encountered a protocol declaration, which can only be defined at the top-level. Something is wrong.
+    assertionFailure("Encountered a protocol. This is a usage error: a single NestableTypeVisitor instance should start walking only over a nestable declaration syntax node")
+    return .skipChildren
+  }
+
+  public override func visit(_ node: ExtensionDeclSyntax) -> SyntaxVisitorContinueKind {
+    // We've encountered an extension declaration, which can only be defined at the top-level. Something is wrong.
+    assertionFailure("Encountered an extension. This is a usage error: a single NestableTypeVisitor instance should start walking only over a nestable declaration syntax node")
+    return .skipChildren
+  }
+
+  // MARK: Private
+
+  private func visitNestableDeclaration<DeclSyntax: NestableDeclSyntax, TypeInfo: NestableTypeInfo>(
```

Comment:
This method has the same structure as the `public override func visit(_ node: *DeclSyntax) -> SyntaxVisitorContinueKind` methods declared in the `ClassVisitor`, `StructVisitor`, and `EnumVisitor`

We use generics here to avoid having unreachable lines of code in this method.